doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,669 @@
+  sync.
+
+#### Cons (cursor object specific)
+
+- Additional metadata in the object is limited to whole-batch data.
+- Wasted processing of empty lists by the host when pages are exhausted
+  i.e. when there are still pages remaining for some inputs but not
+  others.
+
+### Batched paging option 3 - Per-element paging
+
+E.g. Page token
+
+```python
+# Callback taking index of batch element and selected page of results.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+    batch: List[EntityReference],
+    relationship,
+    successCb: SuccessCallback,
+    errorCb,
+    limit: -1,
+    pageTokens: List[AbstractPageToken]
+): -> List[AbstractPageToken]: ...
+```
+
+E.g. Cursor object
+
+```python
+class AbstractCursor:
+    def getPage(self, pageSize: int) -> List[EntityReference]:
+        ...
+    def skip(self, pageSize: int):
+        ...
+
+# Callback taking index of batch element and cursor object.
+SuccessCallback = Callable[[int, AbstractCursor], None]
+
+def relatedEntities(
+    batch: List[EntityReference],
+    relationship,
+    successCb: SuccessCallback,
+    errorCb): ...
+```
+
+#### Pros
+
+- Manager is under no obligation to consider errors in other batch
+  elements when implementing query behaviour, as pagination method need
+  not be synchronized.
+
+#### Pros (cursor object specific)
+
+- Individual cursors can be incremented independently, potentially
+  reducing wasted queries.
+
+#### Cons
+
+- Batching is very difficult to optimize for the manager, especially
+  after the first page.
+- Batch-optimized queries have unpredictable performance profile for
+  the host, given the wide range of potential implementations.
+
+#### Cons (cursor object specific)
+
+- Cannot advance to the next page as a batch operation.
+
+### Outcome
+
+Given the reported difficulty of crafting queries that are both batch
+and paging-aware, we opt to go with Option 1 and split the paging API(s)
+from the batched API(s).
+
+This then leaves only a batch-only signature, plus a decision between
+two remaining paging solutions: (non-batched) page token or cursor
+objects.
+
+## Outcome
+
+We opt for a non-batched cursor object design.
+
+Avoiding conflating batching and pagination simplifies manager plugin
+implementations.
+
+Whilst we may miss out on some theoretical performance, we have been
+unable to find any real-world examples showing this sort of batched-page
+optimization.
+
+A cursor object maximises future extensibility, where new functionality
+can be added without further polluting the `Manager`/`ManagerInterface`
+class. They also naturally capture arbitrary state that the manager
+may wish to store as part of the ongoing query.
+
+Cursor objects also reflect the most common programmatic approach to
+paging through the results of a database query, presenting a familiar
+interface for plugin authors, and mapping well to common back-end SDKs.
+
+## Appendix: Considering manager implementations
```

Comment:
After offline conversation, to reduce confusion re. dictating a particular implementation, moved all examples to this Appendix section in 0b9b211c